File: src/main/kotlin/pers/neige/neigeitems/section/impl/CustomSection.kt
package pers.neige.neigeitems.section.impl

import org.bukkit.OfflinePlayer
import org.bukkit.configuration.ConfigurationSection
import pers.neige.neigeitems.section.SectionParser

// 自定义节点, 具体将在js文件中实现, nashorn会自动解析匹配类型, 因此不用担心重定向问题, 不用将kotlin的function更改为java的function
class CustomSection(
    // 自定义节点ID
    override val id: String,

    // 用于响应私有节点解析
    private val func1: (
        ConfigurationSection,
        HashMap<String, String>?,
        OfflinePlayer?,
        ConfigurationSection?
    ) -> String?,

    // 用于响应即时节点解析
    private val func2: (
        List<String>,
        HashMap<String, String>?,
        OfflinePlayer?,
        ConfigurationSection?
    ) -> String?

) : SectionParser() {

    override fun onRequest(
        data: ConfigurationSection,
        cache: HashMap<String, String>?,
        player: OfflinePlayer?,
        sections: ConfigurationSection?): String? {
        return func1(data, cache, player, sections)
    }

    override fun onRequest(
        args: List<String>,
        cache: HashMap<String, String>?,
        player: OfflinePlayer?,
        sections: ConfigurationSection?): String {
        return func2(args, cache, player, sections) ?: "<$id::${args.joinToString("_")}>"
    }
}

File: src/main/resources/CustomSections/CustomSection.js
// 文件名不重要, 写成啥都行
// main类会自动执行
function main() {
    // 导入相应的类, 这两行看不懂的话直接抄就行
    const SectionManager = Packages.pers.neige.neigeitems.manager.SectionManager.INSTANCE
    const CustomSection = Packages.pers.neige.neigeitems.section.impl.CustomSection

    // 创建自定义节点
    const customSection = new CustomSection(
        // 节点id
        "test",
        /**
         * 用于私有节点解析
         * @param data ConfigurationSection 节点内容
         * @param cache HashMap<String, String>? 解析值缓存
         * @param player OfflinePlayer? 待解析玩家
         * @param sections ConfigurationSection? 节点池
         * @return 解析值
         */
        function(data, cache, player, sections) {
            if (data.contains("values")) return data.getStringList("values").size() + ""
            return null
        },
        /**
         * 用于即时节点解析
         * @param args List<String> 节点参数
         * @param cache HashMap<String, String>? 解析值缓存
         * @param player OfflinePlayer? 待解析玩家
         * @param sections ConfigurationSection? 节点池
         * @return 解析值
         */
        function(args, cache, player, sections) {
            return args.length + ""
        })
    // 节点注册
    SectionManager.loadParser(customSection)
}
